File: BE/src/main/java/com/project/evrental/service/FeedbackService.java
package com.project.evrental.service;

import com.project.evrental.domain.common.BookingStatus;
import com.project.evrental.domain.dto.request.CreateFeedbackRequest;
import com.project.evrental.domain.dto.request.RespondFeedbackRequest;
import com.project.evrental.domain.dto.request.UpdateFeedbackRequest;
import com.project.evrental.domain.dto.response.*;
import com.project.evrental.domain.entity.Booking;
import com.project.evrental.domain.entity.Feedback;
import com.project.evrental.domain.entity.User;
import com.project.evrental.exception.custom.ResourceNotFoundException;
import com.project.evrental.mapper.FeedbackMapper;
import com.project.evrental.repository.BookingRepository;
import com.project.evrental.repository.FeedbackRepository;
import com.project.evrental.repository.UserRepository;
import lombok.AccessLevel;
import lombok.RequiredArgsConstructor;
import lombok.experimental.FieldDefaults;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.oauth2.jwt.Jwt;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Duration;
import java.time.LocalDateTime;
import java.util.*;
import java.util.stream.Collectors;

@Slf4j
@Service
@RequiredArgsConstructor
@FieldDefaults(level = AccessLevel.PRIVATE, makeFinal = true)
public class FeedbackService {

    FeedbackRepository feedbackRepository;
    BookingRepository bookingRepository;
    UserRepository userRepository;

    private static final int EDIT_ALLOWED_DAYS = 7;

    // ==================== RENTER USE CASES ====================

    @Transactional
    public FeedbackResponse createFeedback(CreateFeedbackRequest request) {
        log.info("Creating feedback for booking: {}", request.getBookingId());

        String sub = getCognitoSubFromAuthentication();
        User renter = userRepository.findByCognitoSub(sub)
                .orElseThrow(() -> new ResourceNotFoundException("User not found with cognito sub: " + sub));

        // Validate booking
        Booking booking = bookingRepository.findById(request.getBookingId())
                .orElseThrow(() -> new ResourceNotFoundException("Booking not found with ID: " + request.getBookingId()));

        // Check if booking belongs to renter
        if (!booking.getRenter().getId().equals(renter.getId())) {
            throw new IllegalStateException("You can only create feedback for your own bookings");
        }

        // Check if booking is completed
        if (booking.getStatus() != BookingStatus.COMPLETED) {
            throw new IllegalStateException("Feedback can only be created for completed bookings");
        }

        // Check if feedback already exists
        if (feedbackRepository.existsByBookingId(request.getBookingId())) {
            throw new IllegalStateException("Feedback already exists for this booking");
        }

        Feedback feedback = Feedback.builder()
                .booking(booking)
                .renter(renter)
                .vehicleRating(request.getVehicleRating())
                .stationRating(request.getStationRating())
                .comment(request.getComment())
                .isEdit(false)
                .build();

        Feedback savedFeedback = feedbackRepository.save(feedback);
        log.info("Feedback created with ID: {}", savedFeedback.getId());

        return FeedbackMapper.toResponse(savedFeedback);
    }

    @Transactional(readOnly = true)
    public FeedbackDetailResponse getFeedbackById(UUID feedbackId) {
        log.info("Fetching feedback detail with ID: {}", feedbackId);

        Feedback feedback = feedbackRepository.findById(feedbackId)
                .orElseThrow(() -> new ResourceNotFoundException("Feedback not found with ID: " + feedbackId));

        String sub = getCognitoSubFromAuthentication();
        User currentUser = userRepository.findByCognitoSub(sub)
                .orElseThrow(() -> new ResourceNotFoundException("User not found with cognito sub: " + sub));

        // Check if user has permission to view this feedback
        if (!feedback.getRenter().getId().equals(currentUser.getId()) &&
            !hasAdminOrStaffRole()) {
            throw new IllegalStateException("You don't have permission to view this feedback");
        }

        User respondedByUser = null;
        if (feedback.getRespondedBy() != null) {
            respondedByUser = userRepository.findById(feedback.getRespondedBy()).orElse(null);
        }

        return FeedbackMapper.toDetailResponse(feedback, respondedByUser);
    }

    @Transactional(readOnly = true)
    public FeedbackDetailResponse getFeedbackByBookingId(UUID bookingId) {
        log.info("Fetching feedback for booking: {}", bookingId);

        Feedback feedback = feedbackRepository.findByBookingId(bookingId)
                .orElseThrow(() -> new ResourceNotFoundException("Feedback not found for booking: " + bookingId));

        String sub = getCognitoSubFromAuthentication();
        User currentUser = userRepository.findByCognitoSub(sub)
                .orElseThrow(() -> new ResourceNotFoundException("User not found with cognito sub: " + sub));

        // Check permission
        if (!feedback.getRenter().getId().equals(currentUser.getId()) &&
            !hasAdminOrStaffRole()) {
            throw new IllegalStateException("You don't have permission to view this feedback");
        }

        User respondedByUser = null;
        if (feedback.getRespondedBy() != null) {
            respondedByUser = userRepository.findById(feedback.getRespondedBy()).orElse(null);
        }

        return FeedbackMapper.toDetailResponse(feedback, respondedByUser);
    }

    @Transactional
    public FeedbackResponse updateFeedback(UUID feedbackId, UpdateFeedbackRequest request) {
        log.info("Updating feedback with ID: {}", feedbackId);

        String sub = getCognitoSubFromAuthentication();
        User renter = userRepository.findByCognitoSub(sub)
                .orElseThrow(() -> new ResourceNotFoundException("User not found with cognito sub: " + sub));

        Feedback feedback = feedbackRepository.findById(feedbackId)
                .orElseThrow(() -> new ResourceNotFoundException("Feedback not found with ID: " + feedbackId));

        // Check if feedback belongs to renter
        if (!feedback.getRenter().getId().equals(renter.getId())) {
            throw new IllegalStateException("You can only update your own feedback");
        }

        // Check if within edit time window
        long daysSinceCreation = Duration.between(feedback.getCreatedAt(), LocalDateTime.now()).toDays();
        if (daysSinceCreation > EDIT_ALLOWED_DAYS) {
            throw new IllegalStateException("Feedback can only be edited within " + EDIT_ALLOWED_DAYS + " days of creation");
        }

        boolean hasChanges = false;

        if (request.getVehicleRating() != null) {
            feedback.setVehicleRating(request.getVehicleRating());
            hasChanges = true;
        }
        if (request.getStationRating() != null) {
            feedback.setStationRating(request.getStationRating());
            hasChanges = true;
        }
        if (request.getComment() != null) {
            feedback.setComment(request.getComment());
            hasChanges = true;
        }

        if (hasChanges) {
            feedback.setIsEdit(true);
        }

        Feedback updatedFeedback = feedbackRepository.save(feedback);
        log.info("Feedback updated successfully with ID: {}", feedbackId);

        return FeedbackMapper.toResponse(updatedFeedback);
    }

    @Transactional(readOnly = true)
    public Page<FeedbackResponse> getMyFeedbacks(Pageable pageable) {
        log.info("Fetching all feedbacks for current renter");

        String sub = getCognitoSubFromAuthentication();
        User renter = userRepository.findByCognitoSub(sub)
                .orElseThrow(() -> new ResourceNotFoundException("User not found with cognito sub: " + sub));

        Page<Feedback> feedbacks = feedbackRepository.findByRenterIdOrderByCreatedAtDesc(renter.getId(), pageable);

        return feedbacks.map(FeedbackMapper::toResponse);
    }

    // ==================== ADMIN/STAFF USE CASES ====================

    @Transactional(readOnly = true)
    public Page<FeedbackResponse> getAllFeedbacksWithFilters(
            UUID stationId,
            UUID vehicleId,
            UUID renterId,
            LocalDateTime fromDate,
            LocalDateTime toDate,
            Double minRating,
            Double maxRating,
            Pageable pageable) {
        log.info("Fetching feedbacks with filters");

        Page<Feedback> feedbacks = feedbackRepository.findByFilters(
                stationId, vehicleId, renterId,
                fromDate, toDate, minRating, maxRating, pageable
        );

        return feedbacks.map(FeedbackMapper::toResponse);
    }

    @Transactional
    public FeedbackResponse respondToFeedback(UUID feedbackId, RespondFeedbackRequest request) {
        log.info("Responding to feedback with ID: {}", feedbackId);

        String sub = getCognitoSubFromAuthentication();
        User staff = userRepository.findByCognitoSub(sub)
                .orElseThrow(() -> new ResourceNotFoundException("User not found with cognito sub: " + sub));

        Feedback feedback = feedbackRepository.findById(feedbackId)
                .orElseThrow(() -> new ResourceNotFoundException("Feedback not found with ID: " + feedbackId));

        feedback.setResponse(request.getResponse());
        feedback.setRespondedBy(staff.getId());
        feedback.setRespondedAt(LocalDateTime.now());

        Feedback updatedFeedback = feedbackRepository.save(feedback);
        log.info("Response added to feedback ID: {}", feedbackId);

        return FeedbackMapper.toResponse(updatedFeedback);
    }

    @Transactional
    public FeedbackResponse updateResponse(UUID feedbackId, RespondFeedbackRequest request) {
        log.info("Updating response for feedback with ID: {}", feedbackId);

        String sub = getCognitoSubFromAuthentication();
        User staff = userRepository.findByCognitoSub(sub)
                .orElseThrow(() -> new ResourceNotFoundException("User not found with cognito sub: " + sub));

        Feedback feedback = feedbackRepository.findById(feedbackId)
                .orElseThrow(() -> new ResourceNotFoundException("Feedback not found with ID: " + feedbackId));

        if (feedback.getResponse() == null) {
            throw new IllegalStateException("No response exists to update");
        }

        feedback.setResponse(request.getResponse());
        feedback.setRespondedBy(staff.getId());
        feedback.setRespondedAt(LocalDateTime.now());

        Feedback updatedFeedback = feedbackRepository.save(feedback);
        log.info("Response updated for feedback ID: {}", feedbackId);

        return FeedbackMapper.toResponse(updatedFeedback);
    }

    @Transactional
    public void deleteResponse(UUID feedbackId) {
        log.info("Deleting response for feedback with ID: {}", feedbackId);

        Feedback feedback = feedbackRepository.findById(feedbackId)
                .orElseThrow(() -> new ResourceNotFoundException("Feedback not found with ID: " + feedbackId));

        feedback.setResponse(null);
        feedback.setRespondedBy(null);
        feedback.setRespondedAt(null);

        feedbackRepository.save(feedback);
        log.info("Response deleted for feedback ID: {}", feedbackId);
    }

    @Transactional
    public void deleteFeedback(UUID feedbackId) {
        log.info("Deleting feedback with ID: {}", feedbackId);

        Feedback feedback = feedbackRepository.findById(feedbackId)
                .orElseThrow(() -> new ResourceNotFoundException("Feedback not found with ID: " + feedbackId));

        feedbackRepository.delete(feedback);
        log.info("Feedback deleted successfully with ID: {}", feedbackId);
    }

    // ==================== PUBLIC USE CASES ====================

    @Transactional(readOnly = true)
    public VehicleFeedbackSummaryResponse getVehicleFeedbackSummary(UUID vehicleId, int limit) {
        log.info("Fetching feedback summary for vehicle: {}", vehicleId);

        Double avgRating = feedbackRepository.findAverageVehicleRating(vehicleId);
        Long totalCount = feedbackRepository.countByVehicleId(vehicleId);

        // Get rating distribution
        Long fiveStarCount = feedbackRepository.countByVehicleIdAndRatingRange(vehicleId, 4.5, 6.0);
        Long fourStarCount = feedbackRepository.countByVehicleIdAndRatingRange(vehicleId, 3.5, 4.5);
        Long threeStarCount = feedbackRepository.countByVehicleIdAndRatingRange(vehicleId, 2.5, 3.5);
        Long twoStarCount = feedbackRepository.countByVehicleIdAndRatingRange(vehicleId, 1.5, 2.5);
        Long oneStarCount = feedbackRepository.countByVehicleIdAndRatingRange(vehicleId, 0.0, 1.5);

        // Get recent feedbacks
        Pageable pageable = PageRequest.of(0, limit);
        Page<Feedback> recentFeedbacks = feedbackRepository.findByVehicleId(vehicleId, pageable);

        List<FeedbackResponse> feedbackResponses = recentFeedbacks.getContent().stream()
                .map(FeedbackMapper::toResponse)
                .collect(Collectors.toList());

        return VehicleFeedbackSummaryResponse.builder()
                .vehicleId(vehicleId)
                .averageRating(avgRating != null ? avgRating : 0.0)
                .totalFeedbackCount(totalCount != null ? totalCount : 0L)
                .fiveStarCount(fiveStarCount != null ? fiveStarCount : 0L)
                .fourStarCount(fourStarCount != null ? fourStarCount : 0L)
                .threeStarCount(threeStarCount != null ? threeStarCount : 0L)
                .twoStarCount(twoStarCount != null ? twoStarCount : 0L)
                .oneStarCount(oneStarCount != null ? oneStarCount : 0L)
                .recentFeedbacks(feedbackResponses)
                .build();
    }

    @Transactional(readOnly = true)
    public StationFeedbackSummaryResponse getStationFeedbackSummary(UUID stationId, int limit) {
        log.info("Fetching feedback summary for station: {}", stationId);

        Double avgRating = feedbackRepository.findAverageStationRating(stationId);
        Long totalCount = feedbackRepository.countByStationId(stationId);

        // Get rating distribution
        Long fiveStarCount = feedbackRepository.countByStationIdAndRatingRange(stationId, 4.5, 6.0);
        Long fourStarCount = feedbackRepository.countByStationIdAndRatingRange(stationId, 3.5, 4.5);
        Long threeStarCount = feedbackRepository.countByStationIdAndRatingRange(stationId, 2.5, 3.5);
        Long twoStarCount = feedbackRepository.countByStationIdAndRatingRange(stationId, 1.5, 2.5);
        Long oneStarCount = feedbackRepository.countByStationIdAndRatingRange(stationId, 0.0, 1.5);

        // Get recent feedbacks
        Pageable pageable = PageRequest.of(0, limit);
        Page<Feedback> recentFeedbacks = feedbackRepository.findByStationId(stationId, pageable);

        List<FeedbackResponse> feedbackResponses = recentFeedbacks.getContent().stream()
                .map(FeedbackMapper::toResponse)
                .collect(Collectors.toList());

        return StationFeedbackSummaryResponse.builder()
                .stationId(stationId)
                .averageRating(avgRating != null ? avgRating : 0.0)
                .totalFeedbackCount(totalCount != null ? totalCount : 0L)
                .fiveStarCount(fiveStarCount != null ? fiveStarCount : 0L)
                .fourStarCount(fourStarCount != null ? fourStarCount : 0L)
                .threeStarCount(threeStarCount != null ? threeStarCount : 0L)
                .twoStarCount(twoStarCount != null ? twoStarCount : 0L)
                .oneStarCount(oneStarCount != null ? oneStarCount : 0L)
                .recentFeedbacks(feedbackResponses)
                .build();
    }

    @Transactional(readOnly = true)
    public Page<FeedbackResponse> getPublicVehicleFeedbacks(UUID vehicleId, Pageable pageable) {
        log.info("Fetching public feedbacks for vehicle: {}", vehicleId);

        Page<Feedback> feedbacks = feedbackRepository.findByVehicleId(vehicleId, pageable);
        return feedbacks.map(FeedbackMapper::toResponse);
    }

    @Transactional(readOnly = true)
    public Page<FeedbackResponse> getPublicStationFeedbacks(UUID stationId, Pageable pageable) {
        log.info("Fetching public feedbacks for station: {}", stationId);

        Page<Feedback> feedbacks = feedbackRepository.findByStationId(stationId, pageable);
        return feedbacks.map(FeedbackMapper::toResponse);
    }

    // ==================== STATISTICS ====================

    @Transactional(readOnly = true)
    public FeedbackStatisticsResponse getGlobalStatistics() {
        log.info("Fetching global feedback statistics");

        Long totalCount = feedbackRepository.count();
        Long respondedCount = feedbackRepository.countRespondedFeedbacks();
        Long unrespondedCount = feedbackRepository.countUnrespondedFeedbacks();

        LocalDateTime sevenDaysAgo = LocalDateTime.now().minusDays(7);
        LocalDateTime thirtyDaysAgo = LocalDateTime.now().minusDays(30);

        Long feedbacksLast7Days = feedbackRepository.countFeedbacksSinceDate(sevenDaysAgo);
        Long feedbacksLast30Days = feedbackRepository.countFeedbacksSinceDate(thirtyDaysAgo);

        Double avgResponseTime = feedbackRepository.findAverageResponseTimeInHours();

        // Rating distributions
        List<Object[]> vehicleRatingDist = feedbackRepository.getVehicleRatingDistribution();
        Map<Integer, Long> vehicleRatingMap = new HashMap<>();
        for (Object[] row : vehicleRatingDist) {
            vehicleRatingMap.put((Integer) row[0], (Long) row[1]);
        }

        List<Object[]> stationRatingDist = feedbackRepository.getStationRatingDistribution();
        Map<Integer, Long> stationRatingMap = new HashMap<>();
        for (Object[] row : stationRatingDist) {
            stationRatingMap.put((Integer) row[0], (Long) row[1]);
        }

        // Top rated
        List<Object[]> topVehicles = feedbackRepository.findTopRatedVehicles(PageRequest.of(0, 1));
        Map<String, Object> topVehicle = null;
        if (!topVehicles.isEmpty()) {
            Object[] row = topVehicles.get(0);
            topVehicle = Map.of(
                    "id", row[0],
                    "name", row[1],
                    "averageRating", row[2]
            );
        }

        List<Object[]> topStations = feedbackRepository.findTopRatedStations(PageRequest.of(0, 1));
        Map<String, Object> topStation = null;
        if (!topStations.isEmpty()) {
            Object[] row = topStations.get(0);
            topStation = Map.of(
                    "id", row[0],
                    "name", row[1],
                    "averageRating", row[2]
            );
        }

        return FeedbackStatisticsResponse.builder()
                .totalFeedbackCount(totalCount)
                .respondedCount(respondedCount != null ? respondedCount : 0L)
                .unrespondedCount(unrespondedCount != null ? unrespondedCount : 0L)
                .vehicleRatingDistribution(vehicleRatingMap)
                .stationRatingDistribution(stationRatingMap)
                .topRatedVehicle(topVehicle)
                .topRatedStation(topStation)
                .feedbacksLast7Days(feedbacksLast7Days != null ? feedbacksLast7Days : 0L)
                .feedbacksLast30Days(feedbacksLast30Days != null ? feedbacksLast30Days : 0L)
                .averageResponseTimeHours(avgResponseTime)
                .build();
    }

    // ==================== HELPER METHODS ====================

    private String getEmailFromAuthentication() {
        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
        if (authentication != null && authentication.getPrincipal() instanceof Jwt jwt) {
            return jwt.getClaim("email");
        }
        throw new IllegalStateException("User is not authenticated");
    }

    private String getCognitoSubFromAuthentication() {
        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
        if (authentication != null && authentication.getPrincipal() instanceof Jwt jwt) {
            return jwt.getClaim("sub");
        }
        throw new IllegalStateException("User is not authenticated");
    }

    private boolean hasAdminOrStaffRole() {
        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
        if (authentication != null) {
            return authentication.getAuthorities().stream()
                    .anyMatch(grantedAuthority ->
                            grantedAuthority.getAuthority().equals("ROLE_ADMIN") ||
                            grantedAuthority.getAuthority().equals("ROLE_STAFF"));
        }
        return false;
    }
}

File: FE/aws-project/client/service/feedback/feedbackService.ts
// ==========================================
// Feedback Service
// Handles feedback operations
// ==========================================

import axios from "axios";
import apiClient from "../api/apiClient";
import type {
  FeedbackResponse,
  FeedbackPageResponse,
  CreateFeedbackRequest,
  UpdateFeedbackRequest,
  RespondFeedbackRequest,
  FeedbackDetailResponse,
  FeedbackSummaryResponse,
  FeedbackStatisticsResponse,
  FeedbackAdminFilterParams,
} from "../types/feedback.types";

// API Endpoints for feedbacks
const FEEDBACK_ENDPOINTS = {
  // Renter endpoints
  CREATE: "/feedbacks",
  GET_BY_ID: "/feedbacks/:feedbackId",
  GET_BY_BOOKING: "/feedbacks/booking/:bookingId",
  UPDATE: "/feedbacks/:feedbackId",
  GET_MY_FEEDBACKS: "/feedbacks/my-feedbacks",

  // Admin/Staff endpoints
  GET_ALL_ADMIN: "/feedbacks/admin/all",
  DELETE: "/feedbacks/:feedbackId",
  RESPOND: "/feedbacks/:feedbackId/respond",
  UPDATE_RESPONSE: "/feedbacks/:feedbackId/respond",
  DELETE_RESPONSE: "/feedbacks/:feedbackId/respond",
  STATISTICS: "/feedbacks/statistics",

  // Public endpoints
  GET_BY_VEHICLE: "/feedbacks/vehicle/:vehicleId",
  GET_VEHICLE_SUMMARY: "/feedbacks/vehicle/:vehicleId/summary",
  GET_BY_STATION: "/feedbacks/station/:stationId",
  GET_STATION_SUMMARY: "/feedbacks/station/:stationId/summary",
};

// Create a dedicated axios instance for feedback API that uses idToken
const feedbackApiClient = axios.create({
  baseURL: "/api", // Use Vite proxy
  timeout: 30000,
  headers: {
    "Content-Type": "application/json",
  },
  withCredentials: true,
});

// Request interceptor to add idToken instead of accessToken
feedbackApiClient.interceptors.request.use(
  (config) => {
    const idToken = localStorage.getItem("idToken");
    if (idToken) {
      config.headers.Authorization = `Bearer ${idToken}`;
    }
    console.log("Feedback API request with idToken:", {
      url: config.url,
      method: config.method,
      hasAuth: !!config.headers.Authorization,
    });
    return config;
  },
  (error) => {
    return Promise.reject(error);
  },
);

// Response interceptor for error handling
feedbackApiClient.interceptors.response.use(
  (response) => response,
  (error) => {
    console.error("Feedback API error:", error.response?.data || error.message);
    return Promise.reject(error);
  },
);

class FeedbackService {
  // ==========================================
  // RENTER ENDPOINTS
  // ==========================================

  /**
   * Create a new feedback
   * POST /api/feedbacks
   * @param data - Feedback creation data
   * @returns Created feedback response
   */
  async createFeedback(data: CreateFeedbackRequest): Promise<FeedbackResponse> {
    const response = await feedbackApiClient.post(
      FEEDBACK_ENDPOINTS.CREATE,
      data,
    );
    return response.data?.data || response.data;
  }

  /**
   * Get feedback by ID
   * GET /api/feedbacks/:feedbackId
   * @param feedbackId - Feedback UUID
   * @returns Detailed feedback response
   */
  async getFeedbackById(feedbackId: string): Promise<FeedbackDetailResponse> {
    const url = FEEDBACK_ENDPOINTS.GET_BY_ID.replace(":feedbackId", feedbackId);
    const response = await feedbackApiClient.get(url);
    return response.data?.data || response.data;
  }

  /**
   * Get feedback by booking ID
   * GET /api/feedbacks/booking/:bookingId
   * @param bookingId - Booking UUID
   * @returns Detailed feedback response
   */
  async getFeedbackByBookingId(
    bookingId: string,
  ): Promise<FeedbackDetailResponse> {
    const url = FEEDBACK_ENDPOINTS.GET_BY_BOOKING.replace(
      ":bookingId",
      bookingId,
    );
    const response = await feedbackApiClient.get(url);
    return response.data?.data || response.data;
  }

  /**
   * Update an existing feedback
   * PUT /api/feedbacks/:feedbackId
   * @param feedbackId - Feedback UUID
   * @param data - Feedback update data
   * @returns Updated feedback response
   */
  async updateFeedback(
    feedbackId: string,
    data: UpdateFeedbackRequest,
  ): Promise<FeedbackResponse> {
    const url = FEEDBACK_ENDPOINTS.UPDATE.replace(":feedbackId", feedbackId);
    const response = await feedbackApiClient.put(url, data);
    return response.data?.data || response.data;
  }

  /**
   * Get my feedbacks (current user's feedbacks)
   * GET /api/feedbacks/my-feedbacks
   * @param page - Page number (0-indexed)
   * @param size - Page size
   * @returns Paginated feedback list
   */
  async getMyFeedbacks(
    page: number = 0,
    size: number = 10,
  ): Promise<FeedbackPageResponse> {
    const params = new URLSearchParams({
      page: page.toString(),
      size: size.toString(),
    });

    const response = await feedbackApiClient.get(
      `${FEEDBACK_ENDPOINTS.GET_MY_FEEDBACKS}?${params.toString()}`,
    );
    return response.data?.data || response.data;
  }

  // ==========================================
  // ADMIN/STAFF ENDPOINTS
  // ==========================================

  /**
   * Get all feedbacks with filters (Admin/Staff only)
   * GET /api/feedbacks/admin/all
   * @param filters - Filter parameters
   * @returns Paginated feedback list
   */
  async getAllFeedbacks(
    filters: FeedbackAdminFilterParams = {},
  ): Promise<FeedbackPageResponse> {
    const params = new URLSearchParams();

    if (filters.stationId) params.append("stationId", filters.stationId);
    if (filters.vehicleId) params.append("vehicleId", filters.vehicleId);
    if (filters.renterId) params.append("renterId", filters.renterId);
    if (filters.fromDate) params.append("fromDate", filters.fromDate);
    if (filters.toDate) params.append("toDate", filters.toDate);
    if (filters.minRating !== undefined)
      params.append("minRating", filters.minRating.toString());
    if (filters.maxRating !== undefined)
      params.append("maxRating", filters.maxRating.toString());

    params.append("page", (filters.page ?? 0).toString());
    params.append("size", (filters.size ?? 10).toString());

    const response = await feedbackApiClient.get(
      `${FEEDBACK_ENDPOINTS.GET_ALL_ADMIN}?${params.toString()}`,
    );
    return response.data?.data || response.data;
  }

  /**
   * Delete a feedback (Admin only)
   * DELETE /api/feedbacks/:feedbackId
   * @param feedbackId - Feedback UUID
   */
  async deleteFeedback(feedbackId: string): Promise<void> {
    const url = FEEDBACK_ENDPOINTS.DELETE.replace(":feedbackId", feedbackId);
    await feedbackApiClient.delete(url);
  }

  /**
   * Respond to a feedback (Admin/Staff)
   * POST /api/feedbacks/:feedbackId/respond
   * @param feedbackId - Feedback UUID
   * @param responseText - Response message
   * @returns Updated feedback response
   */
  async respondToFeedback(
    feedbackId: string,
    responseText: string,
  ): Promise<FeedbackResponse> {
    const url = FEEDBACK_ENDPOINTS.RESPOND.replace(":feedbackId", feedbackId);
    const response = await feedbackApiClient.post(url, {
      response: responseText,
    });
    return response.data?.data || response.data;
  }

  /**
   * Update response to a feedback (Admin/Staff)
   * PUT /api/feedbacks/:feedbackId/respond
   * @param feedbackId - Feedback UUID
   * @param responseText - Updated response message
   * @returns Updated feedback response
   */
  async updateResponse(
    feedbackId: string,
    responseText: string,
  ): Promise<FeedbackResponse> {
    const url = FEEDBACK_ENDPOINTS.UPDATE_RESPONSE.replace(
      ":feedbackId",
      feedbackId,
    );
    const response = await feedbackApiClient.put(url, {
      response: responseText,
    });
    return response.data?.data || response.data;
  }

  /**
   * Delete response from a feedback (Admin/Staff)
   * DELETE /api/feedbacks/:feedbackId/respond
   * @param feedbackId - Feedback UUID
   */
  async deleteResponse(feedbackId: string): Promise<void> {
    const url = FEEDBACK_ENDPOINTS.DELETE_RESPONSE.replace(
      ":feedbackId",
      feedbackId,
    );
    await feedbackApiClient.delete(url);
  }

  /**
   * Get global feedback statistics (Admin/Staff)
   * GET /api/feedbacks/statistics
   * @returns Feedback statistics
   */
  async getStatistics(): Promise<FeedbackStatisticsResponse> {
    const response = await feedbackApiClient.get(FEEDBACK_ENDPOINTS.STATISTICS);
    return response.data?.data || response.data;
  }

  // ==========================================
  // PUBLIC ENDPOINTS
  // ==========================================

  /**
   * Get feedbacks by vehicle ID with pagination (Public)
   * GET /api/feedbacks/vehicle/:vehicleId
   * @param vehicleId - Vehicle UUID
   * @param page - Page number (0-indexed)
   * @param size - Page size
   * @returns Paginated feedback list
   */
  async getFeedbacksByVehicle(
    vehicleId: string,
    page: number = 0,
    size: number = 10,
  ): Promise<FeedbackPageResponse> {
    const url = FEEDBACK_ENDPOINTS.GET_BY_VEHICLE.replace(
      ":vehicleId",
      vehicleId,
    );
    const params = new URLSearchParams({
      page: page.toString(),
      size: size.toString(),
    });

    const response = await feedbackApiClient.get(`${url}?${params.toString()}`);
    return response.data?.data || response.data;
  }

  /**
   * Get vehicle feedback summary (Public)
   * GET /api/feedbacks/vehicle/:vehicleId/summary
   * @param vehicleId - Vehicle UUID
   * @param limit - Number of recent feedbacks to include (default 5)
   * @returns Vehicle feedback summary
   */
  async getVehicleFeedbackSummary(
    vehicleId: string,
    limit: number = 5,
  ): Promise<FeedbackSummaryResponse> {
    const url = FEEDBACK_ENDPOINTS.GET_VEHICLE_SUMMARY.replace(
      ":vehicleId",
      vehicleId,
    );
    const params = new URLSearchParams({
      limit: limit.toString(),
    });

    const response = await feedbackApiClient.get(`${url}?${params.toString()}`);
    return response.data?.data || response.data;
  }

  /**
   * Get feedbacks by station ID with pagination (Public)
   * GET /api/feedbacks/station/:stationId
   * @param stationId - Station UUID
   * @param page - Page number (0-indexed)
   * @param size - Page size
   * @returns Paginated feedback list
   */
  async getFeedbacksByStation(
    stationId: string,
    page: number = 0,
    size: number = 10,
  ): Promise<FeedbackPageResponse> {
    const url = FEEDBACK_ENDPOINTS.GET_BY_STATION.replace(
      ":stationId",
      stationId,
    );
    const params = new URLSearchParams({
      page: page.toString(),
      size: size.toString(),
    });

    const response = await feedbackApiClient.get(`${url}?${params.toString()}`);
    return response.data?.data || response.data;
  }

  /**
   * Get station feedback summary (Public)
   * GET /api/feedbacks/station/:stationId/summary
   * @param stationId - Station UUID
   * @param limit - Number of recent feedbacks to include (default 5)
   * @returns Station feedback summary
   */
  async getStationFeedbackSummary(
    stationId: string,
    limit: number = 5,
  ): Promise<FeedbackSummaryResponse> {
    const url = FEEDBACK_ENDPOINTS.GET_STATION_SUMMARY.replace(
      ":stationId",
      stationId,
    );
    const params = new URLSearchParams({
      limit: limit.toString(),
    });

    const response = await feedbackApiClient.get(`${url}?${params.toString()}`);
    return response.data?.data || response.data;
  }
}

// Export singleton instance
const feedbackService = new FeedbackService();
export default feedbackService;
